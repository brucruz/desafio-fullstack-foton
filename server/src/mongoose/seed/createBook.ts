import BookModel from '../../modules/books/BookModel';
import { IBook } from '../../modules/books/IBook';

function getRandomInt(min: number, max: number): number {
  const newMin = Math.ceil(min);
  const newMax = Math.floor(max);
  return Math.floor(Math.random() * (newMax - newMin)) + min;
}

export const startCounters = (): void => {
  global.__COUNTERS__ = {
    books: 0,
  };
};

export const restartCounters = (): void => {
  global.__COUNTERS__ = Object.keys(global.__COUNTERS__).reduce(
    (prev, curr) => ({ ...prev, [curr]: 0 }),
    // eslint-disable-next-line no-undef
    {} as NodeJS.Counters,
  );
};

export const createBook = async (
  args: Pick<IBook, 'title' | 'subtitle' | 'author' | 'cover' | 'description'>,
  randomOrder: boolean | undefined = true,
) => {
  // eslint-disable-next-line no-multi-assign
  const n = global.__COUNTERS__ && (global.__COUNTERS__.books += 1);

  const createdAt = randomOrder
    ? new Date(
        2020,
        getRandomInt(1, 10),
        getRandomInt(1, 28),
        getRandomInt(1, 22),
        getRandomInt(1, 30),
      )
    : new Date(2020, 1, n);

  const book = new BookModel({
    ...args,
    createdAt,
  });

  await book.save();

  return book;
};
