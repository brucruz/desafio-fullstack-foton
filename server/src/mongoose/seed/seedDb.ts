import { initialBooks } from '../initialLoad';
import { createBook } from './createBook';

export async function seedDb() {
  console.log('Started seeding database');

  const insertedBooks = initialBooks.map(async book => {
    const insertedBook = await createBook(book);

    return insertedBook;
  });

  const books = await Promise.all(insertedBooks);

  console.log(`Inserted ${books.length} books to the database`);

  console.log('Seeding ended');
}
