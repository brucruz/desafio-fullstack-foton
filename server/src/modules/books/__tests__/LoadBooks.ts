import { createTestClient } from 'apollo-server-testing';
import { seedDb } from '../../../mongoose/seed/seedDb';
import getTestServer, {
  clearDbAndRestartCounters,
  connectMongoose,
  disconnectMongoose,
} from '../../../test/runTest';

const gql = String.raw;

beforeAll(connectMongoose);
beforeEach(clearDbAndRestartCounters);
afterAll(disconnectMongoose);

const queryGQL = gql`
  query ListBooks($cursor: String, $limit: Int!) {
    books(cursor: $cursor, limit: $limit) {
      books {
        title
        subtitle
        author
        description
        cover
      }
      hasNextPage
      nextCursor
    }
  }
`;

describe('a user', () => {
  it('should be able to list books', async () => {
    await seedDb();

    const variables = {
      limit: 10,
    };

    const testServer = await getTestServer();

    const { query } = createTestClient(testServer());
    const result = await query({
      query: queryGQL,
      variables,
    });

    expect(result.errors).toBeFalsy();
    expect(result.data).toBeTruthy();
    expect(result.data.books.books).toMatchSnapshot();
  });

  it('should be able to load more books', async () => {
    await seedDb();

    const variables = {
      limit: 1,
    };

    const testServer = await getTestServer();

    const { query } = createTestClient(testServer());
    const result = await query({
      query: queryGQL,
      variables,
    });

    const loadMoreResult = await query({
      query: queryGQL,
      variables: {
        limit: 1,
        cursor: result.data.books.nextCursor,
      },
    });

    expect(loadMoreResult.errors).toBeFalsy();
    expect(loadMoreResult.data).toBeTruthy();
    expect(loadMoreResult.data.books.books).toMatchSnapshot();
  });

  it('should be able to query books by their titles', async () => {
    await seedDb();

    const variables = {
      limit: 1,
      query: 'hooked',
    };

    const testServer = await getTestServer();

    const { query } = createTestClient(testServer());
    const result = await query({
      query: queryGQL,
      variables,
    });

    expect(result.errors).toBeFalsy();
    expect(result.data).toBeTruthy();
    expect(result.data.books.books).toMatchSnapshot();
  });
});
