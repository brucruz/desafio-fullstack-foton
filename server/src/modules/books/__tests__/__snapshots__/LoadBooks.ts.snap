// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`a user should be able to list books 1`] = `
Array [
  Object {
    "author": "Gary Keller",
    "cover": "http://books.google.com/books/content?id=rB2ZDQAAQBAJ&printsec=frontcover&img=1&zoom=1&edge=curl&source=gbs_api",
    "description": "• More than 500 appearances on national bestseller lists • #1 Wall Street Journal, New York Times, and USA Today • Won 12 book awards • Translated into 35 languages • Voted Top 100 Business Book of All Time on Goodreads People are using this simple, powerful concept to focus on what matters most in their personal and work lives. Companies are helping their employees be more productive with study groups, training, and coaching. Sales teams are boosting sales. Churches are conducting classes and recommending for their members. By focusing their energy on one thing at a time people are living more rewarding lives by building their careers, strengthening their finances, losing weight and getting in shape, deepening their faith, and nurturing stronger marriages and personal relationships. YOU WANT LESS. You want fewer distractions and less on your plate. The daily barrage of e-mails, texts, tweets, messages, and meetings distract you and stress you out. The simultaneous demands of work and family are taking a toll. And what's the cost? Second-rate work, missed deadlines, smaller paychecks, fewer promotions--and lots of stress. AND YOU WANT MORE. You want more productivity from your work. More income for a better lifestyle. You want more satisfaction from life, and more time for yourself, your family, and your friends. NOW YOU CAN HAVE BOTH — LESS AND MORE. In The ONE Thing, you'll learn to * cut through the clutter * achieve better results in less time * build momentum toward your goal* dial down the stress * overcome that overwhelmed feeling * revive your energy * stay on track * master what matters to you The ONE Thing delivers extraordinary results in every area of your life--work, personal, family, and spiritual. WHAT'S YOUR ONE THING?",
    "subtitle": "The Surprisingly Simple Truth Behind Extraordinary Results",
    "title": "The ONE Thing",
  },
  Object {
    "author": "Nir Eyal",
    "cover": "http://books.google.com/books/content?id=rB2ZDQAAQBAJ&printsec=frontcover&img=1&zoom=1&edge=curl&source=gbs_api",
    "description": "Revised and Updated, Featuring a New Case Study How do successful companies create products people can’t put down? Why do some products capture widespread attention while others flop? What makes us engage with certain products out of sheer habit? Is there a pattern underlying how technologies hook us? Nir Eyal answers these questions (and many more) by explaining the Hook Model—a four-step process embedded into the products of many successful companies to subtly encourage customer behavior. Through consecutive “hook cycles,” these products reach their ultimate goal of bringing users back again and again without depending on costly advertising or aggressive messaging. Hooked is based on Eyal’s years of research, consulting, and practical experience. He wrote the book he wished had been available to him as a start-up founder—not abstract theory, but a how-to guide for building better products. Hooked is written for product managers, designers, marketers, start-up founders, and anyone who seeks to understand how products influence our behavior. Eyal provides readers with: • Practical insights to create user habits that stick. • Actionable steps for building products people love. • Fascinating examples from the iPhone to Twitter, Pinterest to the Bible App, and many other habit-forming products.",
    "subtitle": "How to Build Habit-Forming Products",
    "title": "Hooked",
  },
]
`;

exports[`a user should be able to load more books 1`] = `Array []`;

exports[`a user should be able to query books by their titles 1`] = `
Array [
  Object {
    "author": "Gary Keller",
    "cover": "http://books.google.com/books/content?id=rB2ZDQAAQBAJ&printsec=frontcover&img=1&zoom=1&edge=curl&source=gbs_api",
    "description": "• More than 500 appearances on national bestseller lists • #1 Wall Street Journal, New York Times, and USA Today • Won 12 book awards • Translated into 35 languages • Voted Top 100 Business Book of All Time on Goodreads People are using this simple, powerful concept to focus on what matters most in their personal and work lives. Companies are helping their employees be more productive with study groups, training, and coaching. Sales teams are boosting sales. Churches are conducting classes and recommending for their members. By focusing their energy on one thing at a time people are living more rewarding lives by building their careers, strengthening their finances, losing weight and getting in shape, deepening their faith, and nurturing stronger marriages and personal relationships. YOU WANT LESS. You want fewer distractions and less on your plate. The daily barrage of e-mails, texts, tweets, messages, and meetings distract you and stress you out. The simultaneous demands of work and family are taking a toll. And what's the cost? Second-rate work, missed deadlines, smaller paychecks, fewer promotions--and lots of stress. AND YOU WANT MORE. You want more productivity from your work. More income for a better lifestyle. You want more satisfaction from life, and more time for yourself, your family, and your friends. NOW YOU CAN HAVE BOTH — LESS AND MORE. In The ONE Thing, you'll learn to * cut through the clutter * achieve better results in less time * build momentum toward your goal* dial down the stress * overcome that overwhelmed feeling * revive your energy * stay on track * master what matters to you The ONE Thing delivers extraordinary results in every area of your life--work, personal, family, and spiritual. WHAT'S YOUR ONE THING?",
    "subtitle": "The Surprisingly Simple Truth Behind Extraordinary Results",
    "title": "The ONE Thing",
  },
]
`;
